// @mapbox/mr-ui/form-submit/form-submit.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FormSubmit = function (_React$PureComponent) {
  (0, _inherits3.default)(FormSubmit, _React$PureComponent);

  function FormSubmit() {
    (0, _classCallCheck3.default)(this, FormSubmit);
    return (0, _possibleConstructorReturn3.default)(this, _React$PureComponent.apply(this, arguments));
  }

  FormSubmit.prototype.render = function render() {
    var _props = this.props,
        testId = _props.testId,
        disabled = _props.disabled,
        label = _props.label,
        onSubmit = _props.onSubmit,
        themeButton = _props.themeButton;


    var inputProps = {
      onClick: onSubmit,
      disabled: disabled,
      'data-test': testId
    };

    return _react2.default.createElement(
      'button',
      (0, _extends3.default)({ type: 'button', className: themeButton }, inputProps),
      label
    );
  };

  return FormSubmit;
}(_react2.default.PureComponent);

FormSubmit.propTypes = {
  /** Function called when the button is clicked. */
  onSubmit: _propTypes2.default.func.isRequired,
  /** Submit button label. */
  label: _propTypes2.default.string,
  /** Toggle disabled state of the element. */
  disabled: _propTypes2.default.bool,
  /** Identifier for the purposes of testing. */
  testId: _propTypes2.default.string,
  /** A space seperated list of Assembly class names. */
  themeButton: _propTypes2.default.string
};
FormSubmit.defaultProps = {
  label: 'Submit',
  testId: '',
  disabled: false,
  themeButton: 'btn px24 py12 round-full txt-s'
};
exports.default = FormSubmit;
