// @mapbox/mr-ui/control-textarea/control-textarea.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _omit = require('../utils/omit');

var _omit2 = _interopRequireDefault(_omit);

var _controlLabel = require('../control-label');

var _controlLabel2 = _interopRequireDefault(_controlLabel);

var _controlWrapper = require('../control-wrapper');

var _controlWrapper2 = _interopRequireDefault(_controlWrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propNames = ['id', 'value', 'onChange', 'label', 'noAuto', 'optional', 'aside', 'validationError', 'themeControlTextarea', 'themeControlWrapper', 'themeLabel',
// Passed when used with the Form component
'initialValue', 'validator'];

var ControlTextarea = function (_React$Component) {
  (0, _inherits3.default)(ControlTextarea, _React$Component);

  function ControlTextarea() {
    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ControlTextarea);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.onChange = function (e) {
      return _this.props.onChange(e.target.value, _this.props.id);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  ControlTextarea.prototype.render = function render() {
    var extraProps = (0, _omit2.default)(this.props, propNames);

    var _props = this.props,
        id = _props.id,
        value = _props.value,
        label = _props.label,
        optional = _props.optional,
        aside = _props.aside,
        validationError = _props.validationError,
        noAuto = _props.noAuto,
        themeControlTextarea = _props.themeControlTextarea,
        themeControlWrapper = _props.themeControlWrapper,
        themeLabel = _props.themeLabel;


    var textareaProps = {
      id: id,
      name: id,
      onChange: this.onChange,
      value: value,
      className: themeControlTextarea,
      'aria-required': optional ? false : true,
      'data-test': id + '-textarea'
    };

    if (noAuto) {
      textareaProps.autoCapitalize = 'none';
      textareaProps.autoCorrect = 'off';
      textareaProps.spellCheck = false;
    }

    if (validationError) {
      textareaProps['aria-invalid'] = true;
    }

    return _react2.default.createElement(
      _controlWrapper2.default,
      {
        themeControlWrapper: themeControlWrapper,
        id: id,
        validationError: validationError
      },
      label && _react2.default.createElement(_controlLabel2.default, {
        text: label,
        id: id,
        aside: aside,
        optional: optional,
        themeLabel: themeLabel
      }),
      _react2.default.createElement('textarea', (0, _extends3.default)({}, textareaProps, extraProps))
    );
  };

  return ControlTextarea;
}(_react2.default.Component);

ControlTextarea.propTypes = {
  /** Unique id for this control. Required if you want a `label`. */
  id: _propTypes2.default.string,
  /** The control's value. */
  value: _propTypes2.default.string,
  /**
   * Invoked when the control's value changes.
   * Passed two arguments: the new value, and the `id` prop.
   */
  onChange: _propTypes2.default.func.isRequired,
  /** Label for the control. */
  label: _propTypes2.default.string,
  /** If `true`, autocorrect and spelling suggestions will be disabled. */
  noAuto: _propTypes2.default.bool,
  /** If `true`, the text `(optional)` is appended to the label element. */
  optional: _propTypes2.default.bool,
  /** Additional content inserted alongside the label element. */
  aside: _propTypes2.default.node,
  /** A validation error to display beneath the control. */
  validationError: _propTypes2.default.node,
  /** Classes to apply to the textarea element. */
  themeControlTextarea: _propTypes2.default.string,
  /** Classes to apply to the control wrapper. */
  themeControlWrapper: _propTypes2.default.string,
  /** Classes to apply to the label element. */
  themeLabel: _propTypes2.default.string
};
ControlTextarea.defaultProps = {
  value: '',
  optional: false,
  noAuto: false,
  themeControlTextarea: 'textarea hmin120'
};
exports.default = ControlTextarea;
