// @mapbox/mr-ui/control-radio-set/control-radio-set.js
'use strict';

exports.__esModule = true;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _omit = require('../utils/omit');

var _omit2 = _interopRequireDefault(_omit);

var _controlLegend = require('../control-legend');

var _controlLegend2 = _interopRequireDefault(_controlLegend);

var _controlWrapper = require('../control-wrapper');

var _controlWrapper2 = _interopRequireDefault(_controlWrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ControlRadioSet = function (_React$Component) {
  (0, _inherits3.default)(ControlRadioSet, _React$Component);

  function ControlRadioSet() {
    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ControlRadioSet);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.onChange = function (e) {
      return _this.props.onChange(e.target.value, _this.props.id);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  ControlRadioSet.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        id = _props.id,
        value = _props.value,
        legend = _props.legend,
        options = _props.options,
        optional = _props.optional,
        aside = _props.aside,
        autoFocus = _props.autoFocus,
        validationError = _props.validationError,
        themeRadioContainer = _props.themeRadioContainer,
        themeRadio = _props.themeRadio,
        themeControlWrapper = _props.themeControlWrapper,
        themeLegend = _props.themeLegend;


    var groupProps = {
      id: id,
      role: 'radiogroup',
      'aria-required': optional ? false : true,
      'data-test': id + '-input'
    };

    if (validationError) {
      groupProps['aria-invalid'] = true;
    }

    var renderOptions = function renderOptions(d) {
      var extraProps = (0, _omit2.default)(d, ['value', 'label']);
      return _react2.default.createElement(
        'label',
        {
          key: d.value,
          className: 'radio-container ' + themeRadioContainer
        },
        _react2.default.createElement('input', (0, _extends3.default)({
          value: d.value,
          checked: d.value === value,
          autoFocus: autoFocus && d.value === value,
          onChange: _this2.onChange,
          name: id,
          type: 'radio'
        }, extraProps)),
        _react2.default.createElement('div', { className: themeRadio + ' radio' }),
        d.label
      );
    };

    return _react2.default.createElement(
      _controlWrapper2.default,
      {
        themeControlWrapper: themeControlWrapper,
        id: id,
        validationError: validationError
      },
      _react2.default.createElement(
        'fieldset',
        null,
        legend && _react2.default.createElement(_controlLegend2.default, {
          text: legend,
          id: id,
          aside: aside,
          optional: optional,
          themeLegend: themeLegend
        }),
        _react2.default.createElement(
          'div',
          groupProps,
          options.map(renderOptions)
        )
      )
    );
  };

  return ControlRadioSet;
}(_react2.default.Component);

ControlRadioSet.propTypes = {
  /** Unique id for this control. Required if you want a `label`. */
  id: _propTypes2.default.string.isRequired,
  /**
   * An array of objects that represent each option of the set.
   * Each object should contain two fields:
   *
   * - `label` which can either be a string or valid JSX
   * - `value` a unique string value
   *
   * Any additional fields added will be passed as attributes to the input
   * element.
   */
  options: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    label: _propTypes2.default.node.isRequired,
    value: _propTypes2.default.string.isRequired
  })).isRequired,
  /**
   * Invoked when the control's value changes.
   * Passed two arguments:
   *
   * - The value. A string matching the `value` field of one of the `options`
   *   props passed.
   * - The `id` prop.
   */
  onChange: _propTypes2.default.func.isRequired,
  /** Legend label for the control. */
  legend: _propTypes2.default.string,
  /** The control's value. Can be an empty string to indicate no value. */
  value: _propTypes2.default.string,
  /** If provided, "(optional)" is appended to the value of the legend element. */
  optional: _propTypes2.default.bool,
  /** Additional content inserted alongside the legend element. */
  aside: _propTypes2.default.node,
  /** When set, auto focus attributes are applied to the input that matches the `value` prop passed. */
  autoFocus: _propTypes2.default.bool,
  /** A validation error to display beneath the control. */
  validationError: _propTypes2.default.node,
  /** Classes to apply to the radio container */
  themeRadioContainer: _propTypes2.default.string,
  /** Classes to apply to the radio element */
  themeRadio: _propTypes2.default.string,
  /** Classes to apply to the control wrapper */
  themeControlWrapper: _propTypes2.default.string,
  /** Classes to apply to the legend element */
  themeLegend: _propTypes2.default.string
};
ControlRadioSet.defaultProps = {
  value: '',
  optional: false,
  autoFocus: false,
  themeRadioContainer: 'txt-s block mb6 flex-parent',
  themeRadio: 'mr6 radio--blue radio--s-label inline-block flex-child'
};
exports.default = ControlRadioSet;
