// @mapbox/mr-ui/control-label/control-label.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ControlLabel = function (_React$PureComponent) {
  (0, _inherits3.default)(ControlLabel, _React$PureComponent);

  function ControlLabel() {
    (0, _classCallCheck3.default)(this, ControlLabel);
    return (0, _possibleConstructorReturn3.default)(this, _React$PureComponent.apply(this, arguments));
  }

  ControlLabel.prototype.render = function render() {
    var _props = this.props,
        text = _props.text,
        id = _props.id,
        aside = _props.aside,
        optional = _props.optional,
        themeLabel = _props.themeLabel;

    var labelClasses = 'inline-block ' + themeLabel;

    return _react2.default.createElement(
      'div',
      null,
      _react2.default.createElement(
        'label',
        { htmlFor: id, className: labelClasses },
        text,
        ' ',
        optional && _react2.default.createElement(
          'span',
          { className: 'txt-normal' },
          '(optional)'
        )
      ),
      aside && _react2.default.createElement(
        'span',
        { className: 'inline-block ml12' },
        aside
      )
    );
  };

  return ControlLabel;
}(_react2.default.PureComponent);

ControlLabel.propTypes = {
  /** Label text */
  text: _propTypes2.default.string.isRequired,
  /** Value should match the identifying id of the input element. */
  id: _propTypes2.default.string.isRequired,
  /** Additional content to provide aligned to the right of the label. */
  aside: _propTypes2.default.node,
  /** If true, label text adds (optional) alongside it. */
  optional: _propTypes2.default.bool,
  /** A space separated list of Assembly class names to override the existing presentation. */
  themeLabel: _propTypes2.default.string
};
ControlLabel.defaultProps = {
  themeLabel: 'txt-s txt-bold mb6'
};
exports.default = ControlLabel;
