// @mapbox/mr-ui/modal/modal-actions.js
'use strict';

exports.__esModule = true;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _button = require('../button');

var _button2 = _interopRequireDefault(_button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ModalActions = function (_React$Component) {
  (0, _inherits3.default)(ModalActions, _React$Component);

  function ModalActions() {
    (0, _classCallCheck3.default)(this, ModalActions);
    return (0, _possibleConstructorReturn3.default)(this, _React$Component.apply(this, arguments));
  }

  ModalActions.prototype.renderSecondary = function renderSecondary() {
    var props = this.props;

    if (!props.primaryAction || !props.secondaryAction) {
      return null;
    }

    return _react2.default.createElement(
      'div',
      { className: 'flex-child flex-child--no-shrink mr12' },
      _react2.default.createElement(
        _button2.default,
        {
          variant: 'discouraging',
          size: 'medium',
          onClick: props.secondaryAction.callback,
          'data-test': 'secondary-modal-action'
        },
        props.secondaryAction.text
      )
    );
  };

  ModalActions.prototype.renderTertiary = function renderTertiary() {
    var props = this.props;

    if (!props.primaryAction || !props.secondaryAction || !props.tertiaryAction) {
      return null;
    }

    return _react2.default.createElement(
      'div',
      { className: 'flex-child flex-child--grow flex-child--no-shrink mr12' },
      _react2.default.createElement(
        'button',
        {
          type: 'button',
          className: 'link link--gray txt-bold txt-s',
          onClick: props.tertiaryAction.callback,
          'data-test': 'tertiary-modal-action'
        },
        props.tertiaryAction.text
      )
    );
  };

  ModalActions.prototype.render = function render() {
    var props = this.props;


    var primaryButtonVariant = props.primaryAction.destructive ? 'destructive' : 'primary';

    return _react2.default.createElement(
      'div',
      { className: 'flex-parent flex-parent--center-main flex-parent--center-cross flex-parent--row-reverse flex-parent--wrap' },
      _react2.default.createElement(
        'div',
        { className: 'flex-child flex-child--no-shrink' },
        _react2.default.createElement(
          _button2.default,
          {
            variant: primaryButtonVariant,
            size: 'medium',
            onClick: props.primaryAction.callback,
            'data-test': 'primary-modal-action'
          },
          props.primaryAction.text
        )
      ),
      this.renderSecondary(),
      this.renderTertiary()
    );
  };

  return ModalActions;
}(_react2.default.Component);

exports.default = ModalActions;


ModalActions.propTypes = {
  primaryAction: _propTypes2.default.shape({
    text: _propTypes2.default.string.isRequired,
    callback: _propTypes2.default.func.isRequired,
    destructive: _propTypes2.default.bool
  }).isRequired,
  secondaryAction: _propTypes2.default.shape({
    text: _propTypes2.default.string.isRequired,
    callback: _propTypes2.default.func.isRequired
  }),
  tertiaryAction: _propTypes2.default.shape({
    text: _propTypes2.default.string.isRequired,
    callback: _propTypes2.default.func.isRequired
  })
};
