// @mapbox/mbx-assembly/lib/custom-variables.js
'use strict';

module.exports = {
  'default-primary-interactive-color-dark': '#314ccd',
  'default-primary-interactive-color': '#4264fb',
  'text-color': '#1f3349',
  'pink-dark': '#b43b71',
  pink: '#ee4e8b',
  'pink-light': '#f8c8da',
  'pink-faint': '#fbe5ee',
  'red-dark': '#ba3b3f',
  red: '#f74e4e',
  'red-light': '#f6b7b7',
  'red-faint': '#fbe5e5',
  'orange-dark': '#ba7334',
  orange: '#f79640',
  'orange-light': '#fbcea6',
  'orange-faint': '#feefe2',
  'yellow-dark': '#a4a62d',
  yellow: '#d9d838',
  'yellow-light': '#FFF5A0',
  'yellow-faint': '#FCFCDF',
  'green-dark': '#269561',
  green: '#33c377',
  'green-light': '#afdec5',
  'green-faint': '#e8f5ee',
  'teal-dark': '#136174',
  teal: '#11b4da',
  'teal-light': '#a4deeb',
  'teal-faint': '#d7f1f6',
  'blue-dark': '#314ccd',
  blue: '#4264fb',
  'blue-light': '#aab7ef',
  'blue-faint': '#edf0fd',
  'purple-dark': '#5a3fc0',
  purple: '#7753eb',
  'purple-light': '#c5b9eb',
  'purple-faint': '#f2effa',
  'gray-dark': '#273d56',
  gray: '#607d9c',
  'gray-light': '#c6d2e1',
  'gray-faint': '#f4f7fb',
  // Halfway between gray and gray-light
  'default-secondary-interactive-color': '#93a7bf',
  // disabled and inactive values match default-secondary-interactive-color but with opacity
  'disabled-primary-interactive-color': 'rgba(147, 167, 191, 0.25)',
  'disabled-secondary-interactive-color': 'rgba(147, 167, 191, 0.1)',
  'inactive-text-color': 'rgba(147, 167, 191, 0.45)',
  // Darken colors are based on the same value as 'black' below
  darken5: 'rgba(31, 51, 73, 0.05)',
  darken10: 'rgba(31, 51, 73, 0.10)',
  darken25: 'rgba(31, 51, 73, 0.25)',
  darken50: 'rgba(31, 51, 73, 0.50)',
  darken75: 'rgba(31, 51, 73, 0.75)',
  black: '#070b10'
};
